core/trino-main/src/main/java/io/trino/server/remotetask/TaskInfoFetcher.java

```diff
@@ -294,7 +298,10 @@ synchronized void updateTaskInfo(TaskInfo newTaskInfo)
 
         if (updated && newValue.taskStatus().getState().isDone()) {
             taskStatusFetcher.updateTaskStatus(newTaskInfo.taskStatus());
-            finalTaskInfo.compareAndSet(Optional.empty(), Optional.of(newValue));
+            boolean finalTaskInfoUpdated = finalTaskInfo.compareAndSet(Optional.empty(), Optional.of(newValue));
```

Comment:
It seems to me that we should just refuse to update `finalTaskInfo` here if `missingSpoolingOutputStats` is true.  Updating it will trigger the listener registered here:
https://github.com/trinodb/trino/blob/27e78631fbd88198be04d62dab343be12a1c5a93/core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java#L1690
which will eventually trigger this line:
https://github.com/trinodb/trino/blob/27e78631fbd88198be04d62dab343be12a1c5a93/core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java#L2810
and that's forbidden when the `spoolingOutputStats` is missing.

Storing a `null` stats here guarantees that the `null` check will fail in `retrieveAndDropSpoolingOutputStats()`.